File: leetcode/C++/insertion-sort-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* insertionSortList(ListNode* head) {
        ListNode *sol=new ListNode(-5001), *t, *p;
        for(ListNode *cur=head; cur;){
            for(t=sol; t; p=t, t=t->next){
                if(t->val>cur->val) break;
            }
            p->next=cur;
            cur=cur->next;
            p->next->next=t;
        }
        return sol->next;
    }
};

File: leetcode/Python/insertion-sort-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def insertionSortList(self, head: ListNode) -> ListNode:
        sol, cur, p, t = ListNode(val=-5001), head, None, None
        while cur:
            t=sol
            while t:
                if t.val > cur.val:
                    break
                p, t= t, t.next
            p.next, cur = cur, cur.next
            p.next.next=t
        return sol.next